source/common/crypto/utility.cc

```diff
@@ -41,6 +43,69 @@ std::vector<uint8_t> Utility::getSha256Hmac(const std::vector<uint8_t>& key,
   return hmac;
 }
 
+VerificationOutput Utility::verifySignature(const absl::string_view& hash, void* ptr,
+                                            const std::vector<uint8_t>& signature,
+                                            const std::vector<uint8_t>& clearText) {
+  // Step 1: get public key
+  auto pubkey = reinterpret_cast<EVP_PKEY*>(ptr);
+
+  // Step 2: initialize EVP_MD_CTX
+  bssl::ScopedEVP_MD_CTX ctx;
+
+  // Step 3: initialize EVP_MD
+  const EVP_MD* md = Utility::getHashFunction(hash);
+
+  if (md == nullptr) {
+    return {false, absl::StrCat(hash, " is not supported.")};
+  }
+
+  // Step 4: initialize EVP_DigestVerify
+  int ok = EVP_DigestVerifyInit(ctx.get(), nullptr, md, nullptr, pubkey);
+  if (!ok) {
+    return {false, "Failed to initialize digest verify."};
+  }
+
+  // Step 5: verify signature
+  ok = EVP_DigestVerify(ctx.get(), signature.data(), signature.size(), clearText.data(),
+                        clearText.size());
+
+  // Step 6: check result
+  if (ok == 1) {
+    return {true, ""};
+  }
+
+  std::string err_msg;
+  absl::StrAppend(&err_msg, "Failed to verify digest. Error code: ", ok);
+  return {false, err_msg};
```

Comment:
nit: `fmt::format` or `absl::StrCat` to construct the `err_msg` for brevity?

Reply:
`absl::StrAppend` is preferred when appending to a string. https://github.com/abseil/abseil-cpp/blob/master/absl/strings/str_cat.h

Reply:
Sure, no problem. While here we're constructing a string. But yeah, I don't have a strong preference here.

Reply:
Hmm you're just concatenating so this should be one line: `return {false, absl::StrCat("Failed to verify digest. Error code: ", ok)}`